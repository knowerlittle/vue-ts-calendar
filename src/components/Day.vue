<template>
  <div
    :class="dateClass"
    :style="style"
  >
    <div
      tabindex="0"
      :class="$style['innerDate']"
      @click="emitClickDate"
      @keydown.enter="emitClickDate"
    >
      <p>
        {{ date.date }}
      </p>
    </div>
  </div>
</template>

<script lang="ts">
  import { Component, Vue, Prop } from "vue-property-decorator"
  import CalendarDate from "@/services/CalendarDate"

  @Component({ name: "Day" })

  export default class Day extends Vue {

    /**
     * Luxon date object
     */
    @Prop({ required: true })
    readonly date!: CalendarDate

    /**
     * First Saturday date of the month, used for calculation of date placement
     */
    @Prop({ required: true })
    readonly firstSaturdayDate !: number

    $style: any

    get dateClass () {
      return `${this.$style.date}
      ${this.date.isDisabled() && this.$style.disabled}
      ${this.date.isStartDate() && this.$style.isStartDate}
      ${this.date.isEndDate() && this.$style.isEndDate}
      ${this.date.isBetween() && this.$style.isBetween}
      `
    }

    get style (): string {
      return `grid-area: ${this.rowLineNumber} / ${this.date.weekDay()}`
    }

    get rowLineNumber (): number {
      return this.date.date > this.firstSaturdayDate
        ? Math.ceil((this.date.date - this.firstSaturdayDate) / 7 ) + 1
        : 1
    }

    emitClickDate () {
      if (this.date.isDisabled()) return

      this.$emit('click', this.date)
    }
  }
</script>

<style lang="scss" module>
  $default-border: 1px solid black;

  .date {
    position: relative;
    padding-bottom: 100%;
    z-index: 2;

    p {
      color: pink;
    }

    &:hover,
    &:focus {
      outline: none;
      cursor: pointer;
      background-color: yellow;
      z-index: 20;

      p {
        color: black;
      }
    }
  }

  .innerDate {
    position: absolute;
    width: 100%;
    height: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .disabled {
    &:hover {
      color: gray;
      background-color: unset;
      cursor: not-allowed;

      p {
        color: gray;
      }
    }

    p {
      color: gray;
    }
  }

  .isBetween {
    margin-top: -1px;
    margin-bottom: -1px;
    border-top: $default-border;
    border-bottom: $default-border;
  }

  .isStartDate {
    margin: -1px 0 -1px -1px;
    border-radius: 100% 0 0 100%;
    border-left: $default-border;
    border-top: $default-border;
    border-bottom: $default-border;
  }

  .isEndDate {
    margin: -1px -1px -1px 0;
    border-radius: 0 100% 100% 0;
    border-top: $default-border;
    border-bottom: $default-border;
    border-right: $default-border;
  }

  .isStartDate.isEndDate {
    margin: -1px;
    border-radius: 100%;
  }
</style>
