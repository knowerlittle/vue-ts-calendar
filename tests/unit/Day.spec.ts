import { shallowMount } from '@vue/test-utils'
import Day from '@/components/Day.vue'

describe('Day', () => {
  const luxonMock = jest.fn()

  const defaultOpts = {
    propsData: {
      date: {
        luxon: luxonMock,
        isDisabled: jest.fn().mockReturnValue(true),
        isBetween: jest.fn().mockReturnValue(true),
        isStartDate: jest.fn().mockReturnValue(true),
        isEndDate: jest.fn().mockReturnValue(true),
        weekDay: jest.fn().mockReturnValue(5),
        day: 1,
        date: 1
      },
      firstSaturdayDate: 1
    }
  }

  const factory = (opts = {}) => shallowMount(Day, {
    ...defaultOpts,
    ...opts
  })

  describe('matches snapshot', () => {
    it('when the cases are true for class names', () => {
      const wrapper = factory()

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('when the cases are false for class names', () => {
      const options = {
        propsData: {
          date: {
            luxon: luxonMock,
            isDisabled: jest.fn().mockReturnValue(false),
            isBetween: jest.fn().mockReturnValue(false),
            isStartDate: jest.fn().mockReturnValue(false),
            isEndDate: jest.fn().mockReturnValue(false),
            weekDay: jest.fn().mockReturnValue(4),
            day: 1,
            date: 1
          },
          firstSaturdayDate: 1
        }
      }
      const wrapper = factory(options)

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('when the calculation for applying the grid-area numbers for the inline styles are correct', () => {
      const options = {
        propsData: {
          date: {
            luxon: luxonMock,
            isDisabled: jest.fn().mockReturnValue(false),
            isBetween: jest.fn().mockReturnValue(false),
            isStartDate: jest.fn().mockReturnValue(false),
            isEndDate: jest.fn().mockReturnValue(false),
            weekDay: jest.fn().mockReturnValue(5),
            day: 3,
            date: 20
          },
          firstSaturdayDate: 4
        }
      }
      const wrapper = factory(options)

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('events', () => {
    it('emits a date event when clicked and date is not disabled', () => {
      const options = {
        propsData: {
          date: {
            isDisabled: jest.fn().mockReturnValue(false),
            isBetween: jest.fn().mockReturnValue(false),
            isStartDate: jest.fn().mockReturnValue(false),
            isEndDate: jest.fn().mockReturnValue(false),
            weekDay: jest.fn().mockReturnValue(2)
          },
          firstSaturdayDate: 1
        }
      }
      const wrapper = factory(options)
      wrapper.find('.innerDate').trigger('click')

      // @ts-ignore
      expect(wrapper.emitted('click')[0][0]).toEqual(options.propsData.date)
    })

    it('does not emits a date event when clicked if date is disabled', () => {
      const options = {
        propsData: {
          date: {
            isDisabled: jest.fn().mockReturnValue(true),
            isBetween: jest.fn().mockReturnValue(false),
            isStartDate: jest.fn().mockReturnValue(false),
            isEndDate: jest.fn().mockReturnValue(false),
            weekDay: jest.fn().mockReturnValue(2)
          },
          firstSaturdayDate: 1
        }
      }
      const wrapper = factory(options)

      wrapper.find('.innerDate').trigger('click')

      expect(wrapper.emitted('click')).toBeFalsy()
    })
  })
})
