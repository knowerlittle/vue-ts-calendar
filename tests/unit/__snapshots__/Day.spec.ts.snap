// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day matches snapshot when the calculation for applying the grid-area numbers for the inline styles are correct 1`] = `
<div class="date
    false
    false
    false
    false
    " style="grid-area: 4 / 5;">
  <div tabindex="0" class="innerDate">
    <p>
      20
    </p>
  </div>
</div>
`;

exports[`Day matches snapshot when the cases are false for class names 1`] = `
<div class="date
    false
    false
    false
    false
    " style="grid-area: 1 / 4;">
  <div tabindex="0" class="innerDate">
    <p>
      1
    </p>
  </div>
</div>
`;

exports[`Day matches snapshot when the cases are true for class names 1`] = `
<div class="date
    disabled
    isStartDate
    isEndDate
    isBetween
    " style="grid-area: 1 / 5;">
  <div tabindex="0" class="innerDate">
    <p>
      1
    </p>
  </div>
</div>
`;
